File: src/main/java/xyz/gnarbot/gnar/commands/games/OverwatchLookupCommand.java
package xyz.gnarbot.gnar.commands.games;

import org.json.JSONException;
import org.json.JSONObject;
import xyz.gnarbot.gnar.handlers.commands.Command;
import xyz.gnarbot.gnar.handlers.commands.CommandExecutor;
import xyz.gnarbot.gnar.utils.Note;
import xyz.gnarbot.gnar.utils.NullableJSON;

import java.awt.*;
import java.util.StringJoiner;

import static xyz.gnarbot.gnar.utils.Utils.jsonFromUrl;

@Command(aliases = {"overwatch", "ow"}, usage = "(Battle Tag) [Region]", description = "Look up Overwatch information" +
        " about a " + "player.")
public class OverwatchLookupCommand extends CommandExecutor
{
    public void execute(Note note, String label, String[] args)
    {
        if (args.length == 0)
        {
            note.replyError("You need to provide a BattleTag.");
            return;
        }
        
        if (!args[0].matches("[a-zA-Z1-9]+#\\d+"))
        {
            note.replyError("You did not enter a valid BattleTag.");
            return;
        }
        
        try
        {
            String region = "us";
            if (args.length > 1)
            {
                if (args[1].equalsIgnoreCase("us")
                        || args[1].equalsIgnoreCase("kr")
                        || args[1].equalsIgnoreCase("eu"))
                {
                    region = args[1].toLowerCase();
                }
            }
            StringJoiner joiner = new StringJoiner("\n");
            Note m = note.reply("Finding that player...");
            JSONObject jso;
            try
            {
                jso = jsonFromUrl("https://owapi.net/api/v3/u/" + args[0].replaceAll("#", "-") + "/stats")
                        .getJSONObject(region);
                if (jso == null)
                {
                    m.deleteMessage().queue();
                    note.edit("Unable to find Overwatch Player `" + args[0] + "`.");
                    return;
                }
            }
            catch (JSONException e)
            {
                m.deleteMessage().queue();
                note.edit("Unable to find Overwatch Player `" + args[0] + "` within that region.");
                return;
            }
            
            String tag = args[0].replaceAll("-", "#");
            joiner.add("Battle Tag: **__[" + tag + "](https://playoverwatch.com/en-gb/career/pc/" + region.toLowerCase()
                    + "/" + tag.replaceAll("#", "-") + ")__**");
            joiner.add("Region: **__[" + region.toUpperCase() + "](http://masteroverwatch.com/leaderboards/pc/" +
                    region.toLowerCase() + "/mode/ranked/category/skillrating)__**");
            
            JSONObject overallStats = new NullableJSON(jso.getJSONObject("stats").toString());
            JSONObject compStats = new NullableJSON(overallStats.getJSONObject("competitive").toString());
            JSONObject compOverallStats = new NullableJSON(compStats.getJSONObject("overall_stats").toString());
            JSONObject compGameStats = new NullableJSON(compStats.getJSONObject("game_stats").toString());
            JSONObject compAvgStats = new NullableJSON(compStats.getJSONObject("average_stats").toString());
            JSONObject quickPStats = new NullableJSON(overallStats.getJSONObject("quickplay").toString());
            JSONObject quickPOverallStats = new NullableJSON(quickPStats.getJSONObject("overall_stats").toString());
            JSONObject quickPGameStats = new NullableJSON(quickPStats.getJSONObject("game_stats").toString());
            JSONObject quickPAvgStats = new NullableJSON(quickPStats.getJSONObject("average_stats").toString());
            
            joiner.add("\n**__General                                                      __**");
            joiner.add("  Level: **[" + (quickPOverallStats.getInt("prestige") * 100 + quickPOverallStats.getInt
                    ("level")) + "]()**");
            joiner.add("\n**__Quick Play                                                 __**");
            joiner.add("  Avg. Elims: **[" + quickPAvgStats.getDouble("eliminations_avg") + "]()**");
            joiner.add("  Avg. Deaths: **[" + quickPAvgStats.getDouble("deaths_avg") + "]()**");
            joiner.add("  Avg. Final Blows: **[" + quickPAvgStats.getDouble("final_blows_avg") + "]()**");
            joiner.add("  Wins: **[" + (quickPOverallStats.getInt("wins")) + "]()**");
            joiner.add("  K/D Ratio: **[" + (quickPGameStats.getDouble("kpd")) + "]()**");
            joiner.add("  Played for: **[" + (quickPGameStats.getInt("time_played")) + " hours]()**");
            
            joiner.add("\n**__Competitive                                              __**");
            
            int rank = compOverallStats.getInt("comprank");
            joiner.add("  Avg. Elims: **[" + compAvgStats.getDouble("eliminations_avg") + "]()**");
            joiner.add("  Avg. Deaths: **[" + compAvgStats.getDouble("deaths_avg") + "]()**");
            joiner.add("  Wins/Draws/Loses: **["
                    + compGameStats.getInt("games_won")  + "]()** | **["
                    + compGameStats.getInt("games_tied") + "]()** | **["
                    + compGameStats.getInt("games_lost") + "]()**");
            joiner.add("  K/D Ratio: **[" + compGameStats.getDouble("kpd") + "]()**");
            joiner.add("  Played for: **[" + compGameStats.getInt("time_played") + " hours]()**");
            
            Color sideColor = null;
            String rankName = null;
            if (rank < 1500)
            {
                sideColor = new Color(150, 90, 56);
                rankName = "Bronze";
            }
            else if (rank >= 1500 && rank < 2000)
            {
                sideColor = new Color(168, 168, 168);
                rankName = "Silver";
            }
            else if (rank >= 2000 && rank < 2500)
            {
                sideColor = new Color(201, 137, 16);
                rankName = "Gold";
            }
            else if (rank >= 2500 && rank < 3000)
            {
                sideColor = new Color(229, 228, 226);
                rankName = "Platinum";
            }
            else if (rank >= 3000 && rank < 3500)
            {
                sideColor = new Color(63, 125, 255);
                rankName = "Diamond";
            }
            else if (rank >= 3500 && rank < 4000)
            {
                sideColor = new Color(255, 184, 12);
                rankName = "Master";
            }
            else if (rank >= 4000)
            {
                sideColor = new Color(238, 180, 255);
                rankName = "Grand Master";
            }
            
            joiner.add("  Comp. Rank: **[:beginner: " + rank + "]() (" + rankName + ")**");
            
            joiner.add("\n**__Overall                                                        __**");
            joiner.add("  Eliminations: **[" + (int) (quickPGameStats.getDouble("eliminations") + compGameStats
                    .getDouble("eliminations")) + "]()**");
            joiner.add("  Medals: **[" + (int) (quickPGameStats.getDouble("medals") + compGameStats.getDouble
                    ("medals")) + "]()**");
            joiner.add("  Total Damage: **[" + (int) (quickPGameStats.getDouble("damage_done") + compGameStats
                    .getDouble("damage_done")) + "]()**");
            joiner.add("  Cards: **[" + (int) (quickPGameStats.getDouble("cards") + compGameStats.getDouble("cards")) + "]()**");
            
            m.deleteMessage().queue();
            
            note.replyEmbedRaw("**Overwatch Stats for " + tag + "**", joiner.toString(), sideColor, quickPOverallStats.getString("avatar"));
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: src/main/java/xyz/gnarbot/gnar/commands/games/OverwatchLookupCommand.kt
package xyz.gnarbot.gnar.commands.games

import com.mashape.unirest.http.Unirest
import xyz.gnarbot.gnar.Bot
import xyz.gnarbot.gnar.handlers.commands.Command
import xyz.gnarbot.gnar.handlers.commands.CommandExecutor
import xyz.gnarbot.gnar.utils.Note
import java.util.StringJoiner

@Command(aliases = arrayOf("overwatch", "ow"), usage = "(us/eu/kr) (Battle Tag)", description = "Look up Overwatch information about a player.")
class OverwatchLookupCommand : CommandExecutor()
{
    override fun execute(message : Note, label : String, args : Array<String>)
    {
        when (args.size)
        {
            0 ->
            {
                message.reply("You need to provide a region. `[us, kr, eu]`")
                return
            }
            1 ->
            {
                message.reply("You need to provide a BattleTag.")
                return
            }
        }
    
        val region = args[0].toLowerCase()
    
        if (region != "us" && region != "eu" && region != "kr")
        {
            message.reply("Invalid region. `[us, kr, eu]`")
            return
        }
        
        if (!args[1].matches("""[a-zA-Z1-9]+#\d+""".toRegex()))
        {
            message.reply("You did not enter a valid BattleTag in the form of `[a-zA-Z1-9]#0000`.")
            return
        }
        
        val tag = args[1].replace("#".toRegex(), "-")
    
        try
        {
            val joiner = StringJoiner("\n")
        
            val response = Unirest.get("https://owapi.net/api/v3/u/$tag/stats")
                    .asJson().body.`object`
        
            if (response == null)
            {
                message.reply("Unable to find Overwatch Player `" + args[1] + "`.")
                return
            }
    
            joiner.add("Battle Tag: [**$tag**](https://playoverwatch.com/en-gb/career/pc/$region/$tag)")
            joiner.add("Region: **${region.toUpperCase()}**")
        
            val regionJ = response.getJSONObject(region)
            
            val stats = regionJ.getJSONObject("stats")
    
            
            val quick = stats.getJSONObject("quickplay")
            val quick_avg = quick.getJSONObject("average_stats")
            val quick_overall = quick.getJSONObject("overall_stats")
            val quick_game = quick.getJSONObject("game_stats")
            
            val comp = stats.getJSONObject("competitive")
            val comp_avg = comp.getJSONObject("average_stats")
            val comp_overall = comp.getJSONObject("overall_stats")
            val comp_game = comp.getJSONObject("game_stats")
            
            val icon = quick_overall.getString("avatar")
    
            
            joiner.add("Level: **${(quick_overall.getInt("prestige") * 100 + quick_overall.getInt("level"))}**")
            
            
            
            
            joiner.add("\n**__Quick Play                                 __**")
            joiner.add("  Avg. Elims: **${quick_avg.getDouble("eliminations_avg")}**")
            joiner.add("  Avg. Deaths: **${quick_avg.getDouble("deaths_avg")}**")
//            joiner.add("  Damage Done: **${quick_avg.getDouble("damage_done_avg")}**")
//            joiner.add("  Healing Done: **${quick_avg.getDouble("healing_done_avg")}**")
//            joiner.add("  Objective Kills: **${quick_avg.getDouble("objective_kills_avg")}**")
            joiner.add("  Wins: **${quick_game.getInt("games_won")}**")
            joiner.add("  K/D Ratio: **${quick_game.getDouble("kpd")}**")
//            joiner.add("  Medals: **${quick_game.getInt("medals")}**")
//            joiner.add("  Cards: **${quick_game.getInt("cards")}**")
            joiner.add("  Played for: **${quick_game.getInt("time_played")} hours**")
            
            joiner.add("\n**__Competitive                               __**")
            joiner.add("  Avg. Elims: **${comp_avg.getDouble("eliminations_avg")}**")
            joiner.add("  Avg. Deaths: **${comp_avg.getDouble("deaths_avg")}**")
//            joiner.add("  Damage Done: **${comp_avg.getDouble("damage_done_avg")}**")
//            joiner.add("  Healing Done: **${comp_avg.getDouble("healing_done_avg")}**")
//            joiner.add("  Objective Kills: **${comp_avg.getDouble("objective_kills_avg")}**")
            joiner.add("  Skill Rating: \uD83D\uDD30 **${comp_overall.get("comprank")}**")
            joiner.add("  Wins/Draws/Loses: **${comp_game.getInt("games_won")}** | **${comp_game.getInt("games_tied")}** | **${comp_game.getInt("games_lost")}**")
            joiner.add("  K/D Ratio: **${comp_game.getDouble("kpd")}**")
//            joiner.add("  Medals: **${comp_game.getInt("medals")}**")
//            joiner.add("  Cards: **${comp_game.getInt("cards")}**")
            joiner.add("  Played for: **${comp_game.getInt("time_played")} hours**")
            
            message.replyEmbed("Overwatch Stats", joiner.toString(), Bot.color, icon)
        
        }
        catch (e : Exception)
        {
            e.printStackTrace()
        }
    
    }
}
